redis/connection.py

```diff
@@ -538,9 +540,10 @@ def __init__(
         self.host = host
         self.port = int(port)
         self.db = db
-        self.username = username
         self.client_name = client_name
-        self.password = password
+        self.credentials_provider = credentials_provider
+        if not self.credentials_provider and (username or password):
```

Comment:
I don't believe this is the correct logic for all use cases.

What are your thoughts on the enterprise pattern where a user connects to a service with an initial set (confirmation of connection basically) and then needs to change authentication methods?

Reply:
This check was added only for backward compatibility with username and password.
If the credentials need to be change I think it should be implemented into the credentials supplier function. 
e.g. 
```
def supplier(init_user, init_pass):
   if (init_user or init_pass) and not (self.username or self.password):
        self.username = init_user
        self.password = init_pass
        return self.username, self.password
    return callSupplier()
```

Another option is to add call_supplier kwargs to the CredentialProvider then create a supplier function that changes this filed after the first use:
```
def supplier_use_init_set():
    call_supplier =  self.kwargs.get("call_supplier", True)
    if call_supplier:
        return callSupplier()
    # use the init set only for the first time
    self.kwargs.update({"call_supplier": True})
    return self.username, self.password
    
cred_provider = CredentialProvider(username="init_user", password="init_pass", call_supplier=False, supplier=supplier_use_init_set)
```
Either way I think this logic should be part of the CredentialProvider, and passing Redis with username and password should only be left for backward compatibility.

Reply:
I can add an example for one of those to the connection examples doc.
WDYT?

Reply:
I can also do:
```
        self.credential_provider = credential_provider
        if username or password:
            if self.credential_provider is not None:
                self.credential_provider.username = username
                self.credential_provider.password = password
            else:
                self.credential_provider = CredentialProvider(username, password)
```
but it won't be specific for the mentioned use case